File: src/main/java/com/oyosite/ticon/specutils/data_components/SpectralUtilitiesDataComponents.java
package com.oyosite.ticon.specutils.data_components;

import com.oyosite.ticon.specutils.*;
import net.minecraft.core.*;
import net.minecraft.core.component.*;
import net.minecraft.core.registries.*;

import java.util.function.*;

public class SpectralUtilitiesDataComponents {

	public static final DataComponentType<BlockPos> LINKED_POSITION = register("linked_position",
			(builder) -> builder.persistent(BlockPos.CODEC).networkSynchronized(BlockPos.STREAM_CODEC));

	private static <T> DataComponentType<T> register(String name, UnaryOperator<DataComponentType.Builder<T>> builder) {
		return Registry.register(
				BuiltInRegistries.DATA_COMPONENT_TYPE,
				SpectralUtilities.id(name),
				((DataComponentType.Builder) builder.apply(DataComponentType.builder())).build()
		);
	}

	public static void register() {
	}
}

File: src/main/kotlin/com/oyosite/ticon/specutils/data_components/SpectralUtilitiesDataComponents.kt
package com.oyosite.ticon.specutils.data_components

import com.oyosite.ticon.specutils.SpectralUtilities.Companion.id
import net.minecraft.core.BlockPos
import net.minecraft.core.Registry
import net.minecraft.core.component.DataComponentType
import net.minecraft.core.registries.BuiltInRegistries
import java.util.function.UnaryOperator

object SpectralUtilitiesDataComponents {
    @JvmField
    val LINKED_POSITION: DataComponentType<BlockPos?> = register(
        "linked_position"
    ) { builder: DataComponentType.Builder<BlockPos?> ->
        builder.persistent(BlockPos.CODEC).networkSynchronized(BlockPos.STREAM_CODEC)
    }

    private fun <T> register(
        name: String,
        builder: UnaryOperator<DataComponentType.Builder<T>>
    ): DataComponentType<T?> {
        return Registry.register(
            BuiltInRegistries.DATA_COMPONENT_TYPE,
            id(name),
            (builder.apply(DataComponentType.builder<T>())).build()
        )
    }

    fun register() {
    }
}
